src/node/in_memory_ext.rs

```diff
@@ -51,22 +44,18 @@ impl<S: ForkSource + std::fmt::Debug + Clone + Send + Sync + 'static> InMemoryNo
     /// # Returns
     /// The new timestamp value for the InMemoryNodeInner.
     pub fn set_next_block_timestamp(&self, timestamp: U64) -> Result<U64> {
-        self.get_inner()
-            .write()
-            .map_err(|err| anyhow!("failed acquiring lock: {:?}", err))
-            .and_then(|mut writer| {
-                let ts = timestamp.as_u64();
-                if ts <= writer.current_timestamp {
-                    Err(anyhow!(
-                        "timestamp ({}) must be greater than current timestamp ({})",
-                        ts,
-                        writer.current_timestamp
-                    ))
-                } else {
-                    writer.current_timestamp = ts - 1;
-                    Ok(timestamp)
-                }
-            })
+        let ts = timestamp.as_u64();
+        let last_timestamp = self.time.last_timestamp();
+        if ts <= last_timestamp {
+            Err(anyhow!(
+                "timestamp ({}) must be greater than current timestamp ({})",
+                ts,
+                last_timestamp
+            ))
+        } else {
+            self.time.set_last_timestamp(ts - 1);
+            Ok(timestamp)
```

Comment:
This change is not atomic, so two parallel requests can have TOCTOU. I don't think it can have noticeable negative effects, but still doesn't feel right design-wise. Maybe make it a part of the interface, e.g. make `set_last_timestamp` fallible?

Reply:
Yeah fair enough. Truth be told I did notice the TOCTOUs but they seemed harmless enough. Thanks for keeping me honest :sweat_smile: 

I made a checked an non-checked version of `set_last_timestamp` as there are two flows that expect it to function differently. HTH